Je/multiple buckets

emord opened the pull request:
https://dimagi-dev.atlassian.net/browse/IIO-194

This is at best a 1/10th done PR, but would like to get some feedback on this approach

1. Default configuration stays the same. You are able to keep all blobs in one bucket that is defined in s3 config.
2. All bucket information will be stored in `BlobMeta` in a new `bucket` column. That column can be null if it uses the default blob bucket defined in the s3 configuration. This column should be ignored if you are using `fsdb`
3. A new setting is added `BUCKET_NAME_FUNCTION` with a default of `None`. It can be defined as a function which takes no arguments. Its return value will be the bucket name to store this in.
I did this because, I believe that we want this to be dependent on datetime, or some function that is tracking how many blobs are in the current bucket instead of anything related to the blob itself. Dates make it pretty easy to ensure it rolls over regularly so I went with that as my example.
4. The bucket that an attachment is stored in is decided when `put` is called. 
5. To get the blob you use `meta = BlobMeta.objects.get(); meta.open()` which will handle retrieving from the correct bucket.
  a. This will need to change on a lot of code. There are lots of stray `get_blob_db().get()` code around, but form and case attachments already [proxy through BlobMeta](https://github.com/dimagi/commcare-hq/blob/05b0f5cc099e7f5ea0bfa1f3994ad55b723b3a3b/corehq/form_processor/backends/sql/dbaccessors.py#L451-L452)

I believe I also need to add bucket to `IcdsFile` which appears to avoid using `BlobMeta` at all. I'll also see what I can do to migrate those attachments over to `BlobMeta` if that makes any migration easier.

Things to do:
- [ ] fix migration to work on partitioned databases
- [ ] remove not null constraint so that hte migration can happen without locking
- [ ] add bucket info to `IcdsFile`
- [ ] iterate over all code that uses blobs and ensure they are using `BlobMeta`
- [ ] add bucket paramaters to all interfaces that interact with the blob db

emord committed:
Add bucket to BlobMeta

emord committed:
Support multiple buckets in get and put for s3db

emord committed:
Add setting to create different buckets when putting blobs

emord committed:
Read from correct bucket when opening in BlobMeta

snopoke commented on settings.py:
```diff
@@ -808,6 +808,10 @@
 
+# Takes no arguments and returns a valid bucket naming scheme
+# ex: BUCKET_NAME_FUNCTION = lambda: datetime.datetime.utcnow().strftime('%Y-%m-%d')
+BUCKET_NAME_FUNCTION = None
```
I think this should be a function path string rather than a function itself. In the actual implementation it might even make sense for this to be a callable class so that it can maintain some state about which buckets have been created and also create the new buckets when necessary.

snopoke commented on corehq/blobs/s3db.py:
```diff
@@ -60,12 +63,19 @@ def record_long_request(duration):
-        s3_bucket = self._s3_bucket(create=True)
+        if settings.BUCKET_NAME_FUNCTION:
+            s3_bucket_name = settings.BUCKET_NAME_FUNCTION()
+            s3_bucket = self._s3_bucket(bucket=s3_bucket_name)  # assumes bucket exists
```
in that case you might as well bypass this function altogether since it's whole purpose is to create buckets if they don't exist. You could just do:
```suggestion
            self.db.Bucket(s3_bucket_name)
```

Alternately if the bucket's are being created outside of the blobdb then you could have the `BUCKET_NAME_FUNCTION` return an actual bucket instead of just the name.

emord replied:
Ah. I meant to include a todo in that comment to make sure that it does create the buckets. 

From your above comment:

> I'm now thinking that it might make more sense to have the blobdb track which bucket's it has seen before 

It sounds like you think this approach is better than having `BUCKET_NAME_CLASS` or something similar create the buckets, correct?

snopoke reviewed (commented):
Generally looks good. Made a few comments which can be summarised in the following design questions:

1. where do buckets get created?
  * either in the blobdb itself or somewhere else (like in the 'bucket name function / class')
2. depending on 1, where do we keep track of whether or not a bucket exists.

In my code level comments I was leaning towards having a separate class handle creating the buckets and the blobdb could just use them but I'm now thinking that it might make more sense to have the blobdb track which bucket's it has seen before (only during writes) so that all the bucket creation code is contained within the blobdb.

millerdev commented on corehq/blobs/s3db.py:
```diff
@@ -76,10 +86,10 @@ def put(self, content, **blob_meta_args):
         return meta
 
-    def get(self, key):
+    def get(self, key, bucket=None):
```
Sad to see this come back. [I just removed it](https://github.com/dimagi/commcare-hq/pull/22634/commits/725265bcc8829a9e010f0115033b3c5e1eeeb6f4#diff-1d9ade19245ac6fef5d1bc183aee240aL68), and it made the API much nicer.

millerdev reviewed (commented):
We were discussing yesterday about how I was unhappy with the overloaded concept of "bucket" in the blob db API. I had forgotten that the overloaded concept was completely removed in #22634 (and a huge amount of work that went before that PR). Now we're adding it again because of an implementation detail of the ICDS blob storage backend. This feels wrong to me.

I do not want every use of the blob db have to deal with buckets again, especially considering that most (all?) client code does not care what bucket the blob is stored in. The "bucket" concept would once again be overloaded and client code would need to become aware of implementation details of the blob storage backend (fsdb would have different bucketing behavior from s3db).

Instead, I'd prefer to see the bucket encoded in the blob `key`. For example, prefix keys with a date string from which the bucket can be derived. The blob db would detect this special key format and interact with the proper bucket, possibly (optionally? for backward compatibility with pre-existing keys that may match the pattern) falling back to the default bucket if the key is not found in the bucket indicated by the key. Key generation would need to be audited and adapted to be configurable based on the environment (env). Other than using a different key generation technique, the blob db client API will not change.

emord commented:
> I had forgotten that the overloaded concept was completely removed

I think that this was removed from corehq/blobs aside from some legacy code that can now be removed (some old migrations and #24555). 

References to "bucket" still exists in other places such as https://github.com/dimagi/commcare-hq/blob/master/corehq/form_processor/models.py#L1095. xform attachment sql looks as though it can be deleted, I can't tell with case attachments.

> Instead, I'd prefer to see the bucket encoded in the blob key

This was one of the first things I was thinking about which is likely what caused me to find the bucket/key references in old migration code first. When I saw that, I was concerned about the complexity of ensuring that all current blobs could be handled well. After your comment, I am now thinking of creating a new blobdb backend and doing something such as (keeping this example as brief as possible while explaing the thought):

```python
class MultiBucketS3BlobDb:
    def _parse_bucket_and_key(key):
        return key.split('/'), '/'.join(key.split('/')[1:])

    def put(key):
        bucket, key = _parse_bucket_and_key(key)
        return Bucket(bucket, create=True).put(key) # always put in the new bucket

    def get(orig_key):
        bucket, key = _parse_bucket_and_key(orig_key)
        try:
           return Bucket(bucket).get(key) # optimistically attempt from parsed bucket
        except:
           return DefaultBucket().get(orig_key) # for older keys, try to get from the default bucket
```

1. Does creating a different backend sound like a good idea, or should we include this sort of workflow in the current S3Db?
2. For key generation, I think there may be three options:
    a. The client is responsible for key generation and it should include the namespacing for a bucket. I thought that this originally sounded like what you were suggesting, but I think it also violates the spirit of your statement "do not want every use of the blob db have to deal with buckets again".
    b. The client continues creating a key as it is now. When blob_db gets a put request, it adds a bucket to the key if its a `MultiBucketS3BlobDb` (this could be configured per env in the blob db's settings). The put returns a `BlobMeta` instance where this new key is saved. I think this option is more in the spirit of not having client code concerned with buckets, and only with a single key.
    c. The key continues to be generated as a uuid. The blobdb has a hashing function to determine what bucket to look for that key in. I think the only way I would feel comfortable implementing this is if we saved the bucket to the `BlobMeta` object, as that hash function would never be able to change or any changes would need to be backwards compatible which seems difficult to maintain.

I think for either a or b some client code will need to change to know about buckets (either directly or indirectly) because currently we are currently assuming key can be a uuid. For example soil uses keys only based on the uuid generated when some one creates a task https://github.com/dimagi/commcare-hq/blob/2e8929876d01f47dd1f624110965ad2dd4e240a9/corehq/ex-submodules/soil/__init__.py#L58. It will be strange to include a bucket there or use a blob utility to create a bucket id, so we'll need to add a table to translate between soil IDs and blob keys (arguably that may not be a terrible change long term, but this will need to happen on other models as well). 

Anyways wanted to lay out the options I've been thinking about. I was hoping to get a bit further on this today and be able to come back with a prototype, but was not able to. For now I'm leaning towards option a for key generation with the new `MultiBucketS3BlobDb` class, but would like to hear your thoughts on what you think will be the best long term

snopoke commented:
Personally I prefer having the bucket as an explicit arg rather than 'hiding' it in the key. There are also other things to consider apart from just the blobdb API such as being able to tell how many objects are in a bucket (for example). If the bucket was part of the key then we'd need to parse the key to get it back which would make getting any kind of metrics very hard. I'm not prepared to argue this point too far but I think it is worth considering.

I do see the challenge of shielding the clients from any knowledge of the bucket but I don't think there's a way around that. Either the client specifies the bucket itself or the the blobdb tells it what bucket was used and the client must store that somewhere for future use (either as the bucket name or as part of the key). The only other option is for the blobdb to look up the bucket from the BlobMeta but this would require passing the parent ID.

Either way, there's no way to completely shield the client from knowledge of the bucket. There seem to be pro's and cons for all approaches and it's regrettable we need to do this at all but that can't be helped at this point.

millerdev commented:
I'd like to propose we take this to voice to iron out some higher-level concerns before diving into more implementation details.

emord commented:
After going to voice, we have decided on not taking this approach and will be pushing back on the object storage implementation. I have left as much context as I could think of (including a possible new solution we talked about on the call in [the original ticket](https://dimagi-dev.atlassian.net/browse/IIO-194))
